snowfakery/standard_plugins/datasets.py

```diff
@@ -0,0 +1,220 @@
+from pathlib import Path
+from csv import DictReader
+from contextlib import contextmanager
+import os
+from random import shuffle
+
+from sqlalchemy import create_engine, MetaData
+from sqlalchemy.sql.expression import func, select
+from sqlalchemy.sql.elements import quoted_name
+
+import yaml
+from yaml.representer import Representer
+
+from snowfakery.plugins import SnowfakeryPlugin, PluginResult
+
+
+def _open_db(db_url):
+    "Internal function for opening the database up."
+    engine = create_engine(db_url)
+    metadata = MetaData(bind=engine)
+    metadata.reflect(views=True)
+    return engine, metadata
+
+
+def sql_dataset(db_url: str, tablename: str = None, mode="linear"):
+    "Open the right SQL Dataset iterator based on the params"
+    assert db_url
+    engine, metadata = _open_db(db_url)
+    tables = {
+        name: value
+        for name, value in metadata.tables.items()
+        if not name.startswith("sqlite")
+    }
+    if tablename:
+        table = tables.get(tablename)
+        if table is None:
+            raise AttributeError(f"Cannot find table: {tablename}")
+    elif len(tables) == 0:
+        raise Exception("Database does not exist or has no tables in it")
+    elif len(tables) == 1:
+        table = next(iter(tables.values()))
+    elif len(tables) > 1:
+        raise Exception(
+            f"Database has multiple tables in it and none was selected: {metadata.tables.keys()}"
+        )
+    if mode == "linear":
+        return SQLDatasetLinearIterator(engine, table)
+    elif mode == "shuffle":
+        return SQLDatasetRandomPermutationIterator(engine, table)
+    else:
+        raise NotImplementedError()
+
+
+class DatasetIteratorBase:
+    """Base class for Dataset Iterators
+
+    Subclasses should implement 'self.restart' which puts an iterator into 'self.results'
+    """
+
+    def __next__(self):
+        try:
+            return next(self.results)
+        except StopIteration:
+            self.restart()
+
+    def start(self):
+        "Initialize the iterator in self.results."
+        raise NotImplementedError()
+
+    def restart(self):
+        "Restart the iterator by assigning to self.results"
+        self.start()
+
+    def close(self):
+        "Subclasses should implement this if they need to clean up resources"
+        pass
+
+
+class SQLDatasetIterator(DatasetIteratorBase):
+    def __init__(self, engine, table):
+        self.connection = engine.connect()
+        self.table = table
+        self.start()
+
+    def start(self):
+        self.results = (
+            PluginResult(dict(row)) for row in self.connection.execute(self.query())
+        )
```

Comment:
If you anticipate loading large datasets you might want to use https://docs.sqlalchemy.org/en/13/orm/query.html?highlight=yield_per#sqlalchemy.orm.query.Query.yield_per to limit the size of the working set of data fetched from the db

Reply:
You need that if you're using the ORM but I'm not.

The execute() function returns a ResultProxy:

https://docs.sqlalchemy.org/en/13/core/connections.html#sqlalchemy.engine.ResultProxy

Which is "A facade around a DBAPI cursor object."

In general, I suspect we use ORM functions and "the ORM" too much in CumulusCI, by the way. Now that I've worked with SQL Alchemy for 18 months I have a better intuition of when to use the ORM or not. I suspect all of the bulk loading stuff would be simpler if we didn't use any ORM functions but I'm not proposing a big-bang rewrite.